server/services/rules-service.ts

```diff
@@ -129,13 +130,200 @@ export class RulesService {
+    const agent = agentsService.getSourceAgent(agentId);
+    if (!agent) {
+      logger.error(`Cannot run rule for unknown source agent: ${agentId}`);
+      return { status: 'failed' };
+    }
+    logger.info(`Executing task via agent ${agentId}`);
+    return agent.executeTask(task);
+  }
+
+  /**
+   * Executes a set of target agent tasks.
+   * @param {TargetAgentTask[]} targetTasks the tasks to execute
+   * @returns {Promise<boolean>} completes when all task have been executed
+   */
+  async executeTargetAgentTasks(
+    targetTasks: TargetAgentTask[]
+  ): Promise<boolean> {
+    let success = true;
+    const involvedAgentIds = new Set(
+      ...targetTasks.map((task) => task.agentId)
+    );
+    const involvedAgents: TargetAgent[] = [];
+
+    [...involvedAgentIds].forEach((agentId) => {
+      const agent = agentsService.getTargetAgent(agentId);
+      if (!agent) {
+        logger.error(
+          `Received target agent task for unknown agent: ${agentId}`
+        );
+        return;
+      } else {
+        involvedAgents.push(agent);
+      }
+    });
+
+    involvedAgents.forEach(async (agent) => {
+      const tasksForAgent = targetTasks.filter(
+        (task) => (task.agentId = agent.id)
+      );
+      const results = await agent.executeTasks(tasksForAgent);
+      results.forEach((result) => {
+        if (result.status !== 'success') {
+          success = false;
+          logger.error(`Target agent task did not succeed: ${agent.id}`);
+        }
+      });
+    });
+
+    return success;
+  }
+
+  /**
+   * Determines if a rule is up for execution.
+   * @param {Rule} rule the rule to inspect
+   * @returns {boolean} true if the rule needs execution, false otherwise
    */
-  evaluateRules(rules: Rule[], data: SourceAgentData): RuleEvaluationResult[] {
-    return rules.map((rule) => this.evaluateRule(rule, data));
+  private isRuleInNeedOfExecution(rule: Rule) {
+    // For first time executions lastExecution will not be set
+    if (rule.latestStatus?.lastExecution) {
+      logger.debug(
+        `Rule [${rule.id}] last execution date: ${rule.latestStatus.lastExecution}`
+      );
+      const nextExecutionDate = dateAdd(rule.latestStatus.lastExecution, {
+        minutes: rule.executionInterval,
+      });
+      logger.debug(
+        `Rule [${rule.id}] next execution date: ${nextExecutionDate}`
+      );
+      const needsExecution = isDatePast(nextExecutionDate);
+      if (needsExecution) {
+        logger.debug(`Rule [${rule.id}] needs execution.`);
+      } else {
+        logger.debug(`Rule [${rule.id}] does not need execution.`);
+      }
+
+      return needsExecution;
+    } else {
+      logger.debug(`Rule [${rule.id}] has never run and needs execution.`);
+      return true;
+    }
+  }
+
+  /**
+   * Run all rules.
+   */
+  async runAll() {
+    const rules = await rulesService.listRules();
+    const users = collectionService.users;
+
+    for (const rule of rules) {
```

Comment:
I think you spoke about it at some point, but why isn't it a filter() / map () chain?

Reply:
Filter/map do not handle async/await. Then you'd be map/filtering Promise arrays, which could reduce readability.